Implement TimeStamping feature

hidasw opened the pull request:
Also add some file for parsing asn.1 data, and logging (debug only).
should merged with main tcpdf.php but its not meet oop standar yet.
still use procedural style.

hidasw committed:
Update tcpdf.php

implement Timestamping feature

hidasw committed:
Create asn1_function_tcpdf.php

hidasw committed:
Create asn1_parser_tcpdf.php

hidasw committed:
Create functionLog_tcpdf.php

hidasw committed:
Create Local Root CA.crt

hidasw committed:
Create tcpdf test.pem

hidasw committed:
Create index.php

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/tecnickcom/TCPDF?pullRequest=617) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/tecnickcom/TCPDF?pullRequest=617) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/tecnickcom/TCPDF?pullRequest=617) it.</sub>

hidasw commented on tcpdf.php:
```diff
@@ -7692,12 +7693,14 @@ public function Output($name='doc.pdf', $dest='I') {
-			// add TSA timestamp to signature
-			$signature = $this->applyTSA($signature);
 			// convert signature to hex
 			$signature = current(unpack('H*', $signature));
```
should before $this->applyTSA($signature)

hidasw commented on tcpdf.php:
```diff
@@ -1274,7 +1274,8 @@ class TCPDF {
 	 * @protected
 	 * @since 4.6.005 (2009-04-24)
 	 */
-	protected $signature_max_length = 11742;
```
i''m increasing this size to provide enough space to place tsa response data. and sometimes need more, depend on tsa response data size.

hidasw committed:
Update tcpdf.php

add Content-Type and User-Agent header for tsa request. Some server only accept application/timestamp-query

hidasw committed:
Update index.php

williamdes commented on include/functionLog_tcpdf.php:
```diff
@@ -0,0 +1,49 @@
+<?php
+// create 11:16 AM 10/17/2011
+function tsaLog($str, $type = 'i', $nl=true) { // 11:16 AM 10/17/2011
```
I think this file and function should be removed from this pull-request

hidasw replied:
yes it for debug only

williamdes commented on tcpdf.php:
```diff
@@ -13675,6 +13678,91 @@ protected function applyTSA($signature) {
+						'Content-Type: application/timestamp-query',
+						'User-Agent: TCPDF'
+				curl_setopt($ch, CURLOPT_POSTFIELDS, $raw_data);
+				$tsResponse = curl_exec($ch);
+				if($tsResponse != false) {
+					$tsaLog .= __FILE__." line:(".__LINE__."). tsRequest is sent.\n";
+				} else {
+					tsaLog("$tsaLog".__FILE__." line:(".__LINE__."). can't send tsRequest, Timestamp failed!\n",'w');
+				}
+				//parse ts response
+				$hexTs = bin2hex($tsResponse);
+				$tsparse = asn1parse($hexTs);
+
+				$tsparse0 = asn1parse($tsparse[0][1]);
+				if(count($tsparse0) > 1) { //Remove response status data, only take timeStampToken
+					$timeStamp = seq($tsparse0[1][1]);
+				} else {
+					$timeStamp = seq($tsparse0[0][1]);
+				}
+
+				//Add timestamp to TCPDF Signature
+				$timeStamp = seq("060B2A864886F70D010910020E".set($timeStamp));
+				$pkcs7 = int($pa1[0][1]).seq($pa1[1][1]).seq($pa1[2][1]).explicit(0, $pa1[3][1]).seq($pa1[4][1]).oct($pa1[5][1]);
+				$time = seq($pkcs7.explicit(1,$timeStamp));
+				$aa=seq(int(1). set($p3[1][1]).seq($p3[2][1]).explicit(0, $p3[3][1]).set($time));
+				$hdaSignature = seq("06092A864886F70D010702".explicit(0,($aa)))."0000";
+				
+				$signature = $hdaSignature;
+				// $tsaLog .= $signature;
+				tsaLog("$tsaLog".__FILE__." line:(".__LINE__."). Timestamp Success.\n");
+			} else {
+				$tsaLog .= __FILE__." line:(".__LINE__."). Curl was not loaded, trying to load it...\n";
+				if(@dl('php_curl.dll')) {
+					$tsaLog .= __FILE__." line:(".__LINE__."). Curl successfully Loaded.\n";
+				} else {
+					tsaLog("$tsaLog\n".__FILE__." line:(".__LINE__."). Curl failed to load. Timestamping failed!", 'w');
+				}
+			}
+		}
+		// end timestamping
```
can you do some code cleanup and formatting please ?
It's really hard to read

williamdes commented on tcpdf.php:
```diff
@@ -13675,6 +13678,94 @@ protected function applyTSA($signature) {
+				curl_setopt($ch, CURLOPT_RETURNTRANSFER, 1);
+				curl_setopt($ch, CURLOPT_POST, 1);
+				curl_setopt($ch, CURLOPT_HTTPHEADER, 
+					    	[
+						'Content-Type: application/timestamp-query',
+						'User-Agent: TCPDF'
+						]
```
```suggestion
					    	array(
						'Content-Type: application/timestamp-query',
						'User-Agent: TCPDF'
						)
```
PHP 5.3 is still supported..

hidasw committed:
reformatting and cleanup

also merge asn.1 function to one file,
remove useless log function file.
tested working trusted timestamp http://timestamp.apple.com/ts01

williamdes commented on tcpdf.php:
```diff
@@ -13675,7 +13680,127 @@ protected function applyTSA($signature) {
+			}
+
+			// Add timestamp in TCPDF Signature
+			// Create timestamp pkcs#7 data
+			$TimeStampToken = seq(
+								"060B2A864886F70D010910020E".	// OBJ_id_smime_aa_timeStampToken
+								set(
+									seq(
+										$TSTInfo // TSTInfo
+										)
+									)	
+								);
+			
+			$time = seq(
+						$pkcs7signerInfos[0][1].
+						explicit(1,
+								$TimeStampToken
+								)
+						);
+			
+			$pkcs7contentSignedData=seq(
+										int(1).	// version
+										set($pkcs7SignedData[1][1]).	// digestAlgorithms
+										seq($pkcs7SignedData[2][1]).	// contentInfo
+										explicit(0,
+												$pkcs7SignedData[3][1]
+												).	// certificates [0] IMPLICIT ExtendedCertificatesAndCertificates
+										set(
+											$time
+											)
+										);
+			$pkcs7ContentInfo = seq(
+									"06092A864886F70D010702".	//	ContentType OBJ_pkcs7_signed
+									explicit(0,($pkcs7contentSignedData))	//	content
+									).
+									// "0000"; // sometime needed for backward compatibility
+									"";
+			
+			$signature = $pkcs7ContentInfo;
+		}
```
can you indent this with spaces and move all this code into a protected function you would call here to fetch the signature ?
That would avoid adding that much code into this function

Also, I think you could keep the actual value of signature_max_length and only increase it (if the value < what you need) when calling the signature process ?
Would that keep other pdf documents the same as before (the ones without a signature)

hidasw replied:
i think signature_max_length value have to be reserved before signing process, thats why it not set to 0. will padded by zero after signing. this value affect only signed file size to several kilobytes. Tried to lower this value around 14k and still done with apple tsa.
want move to object oriented but not at the moment. maybe to closed this pull request while i'll periodically update script.

williamdes replied:
Please do not close this pull request
If you invite me on your repository I may clean up the implementation for you

hidasw replied:
ok.

williamdes replied:
Can I start pushing work or do you have more to push first?

hidasw replied:
wait for awhile. i'm updating code to classes.

williamdes replied:
okay, let me know

hidasw replied:
I have done, please review and make some corrections if needed.
I've minimized the script, considering most of the general purpose asn.1 encoders/decoders use complex libraries.

williamdes commented on include/tcpdf_asn1.min.php:
```diff
@@ -0,0 +1,116 @@
+<?php
+// ASN.1 Parser start 21:31 Sore Kamis 26 Maret 2009
+// ASN.1 Parser at 22:10 Sore Kamis 26 Maret 2009 Telah jadi utk standar asn.1
+// 
+// 06:40 Esuk Jumat 27 Maret 2009 ASN.1 Parser kesulitan dlm memecahkan explicit > 9
+
+// 11:18 Esuk Jumat 27 Maret 2009 parse explicit:xx mulai dipecahkan. kemungkinan tlh jadi
+// 17:51 Sore Jumat 27 Maret 2009 memecahkan explicit sampai 2097151 (65536 * 32) kurang 1
+
+// 20:04 Sore Jumat 27 Maret 2009 ASN.1 Parser tlh jadi. Congratulation....
+// 12:15 Sore 16/05/2023
```
all this file seems quite old, what is it source ?
what is the license of this code ?

hidasw replied:
no, it useless. just progress mark. i'm forget to remove some comment line.

hidasw committed:
Update tcpdf_asn1.min.php

hidasw committed:
Update tcpdf.php

williamdes commented on include/tcpdf_asn1.min.php:
```diff
@@ -0,0 +1,111 @@
+  return $ret;
+}
+
+function SEQ($hex)  {
+  $ret = "30".asn1_header($hex).$hex;
+  return $ret;
+}
+function OCT($hex)  {
+  $ret = "04".asn1_header($hex).$hex;
+  return $ret;
+}
+function INT($int)  {
+  if(strlen($int)%2 != 0)  {
+    $int = "0$int";
+  }
+  $int = "$int";
+  $ret = "02".asn1_header($int).$int;
+  return $ret;
+}
+function SET($hex)  {
+  $ret = "31".asn1_header($hex).$hex;
+  return $ret;
+}
+//function EXPLICIT($num="0", $hex)  {
+function EXPLICIT($num, $hex)  {
+  $ret = "a$num".asn1_header($hex).$hex;
+  return $ret;
+}
```
I think you should rename the function starting with `asn1_` so it had less changes to collide with existing functions at the user level

Or prefix them with `tcpdf_`

hidasw replied:
altough renaming that, the procedural style function too exposed indeed, plan to built a class for that but dont know how to do it. cause its so impractice when repeatedly write "new" or "$this->".

williamdes replied:
use `static` 

```php
class Foo {
public static function doFoo() {}
}

Foo::doFoo();
```

Should be way easier

hidasw replied:
I think so, but I need to simplify and compacting the script first.

hidasw replied:
@williamdes, pull request #622 whats that mean? (i'm github newbie :) )

hidasw committed:
major update

migrate procedural to class

williamdes commented on tcpdf.php:
```diff
@@ -13642,6 +13644,8 @@ protected function getSignatureAppearanceArray($x=0, $y=0, $w=0, $h=0, $page=-1,
 	 * @since 6.0.090 (2014-06-16)
 	 */
+	// other options suggested to be implement: reqPolicy, nonce, certReq, extensions 
+	// and option to abort signing if timestamping failed and LTV enable (embed crl and or ocsp revocation info)
```
@nicolaasuni what minimal set of changes would be accepted ?

Some of it could go into a new library ?
Would everything be accepted ?

hidasw replied:
> https://github.com/dealfonso/sapp/blob/main/pdfsign.php

but it for signing, which already implemented here

williamdes replied:
Thanks for the feedback, @hidasw what do you think ?

hidasw replied:
We using openssl for cryptographic functions (encrypt/decrypt). For ocsp/ts/crl operations, php openssl does not have those functions natively, even for asn.1 parsing. So it requires its own independent function or using a library like phpseclib.

williamdes replied:
Can you at least copy the functions from phpseclib so we can know they are safely implemented ?

hidasw replied:
@williamdes 
Unfortunately in phpseclib there is no function for ocsp, ts and crl. There is asn.1 function but it looks too complicated with namespace and many files included. You can compare with my script.
If you are looking for a library for ocsp, crl and ts functions, it is definitely related to the asn.1 function contained in the library. It will be difficult to separate them.

williamdes commented on tcpdf.php:
```diff
@@ -13670,11 +13674,31 @@ public function setTimeStamp($tsa_host='', $tsa_username='', $tsa_password='', $
 	 * @author Richard Stockinger
 	 * @since 6.0.090 (2014-06-16)
 	 */
+	
+	/**
+	 * Applying TSA for a timestamp.
+	 * @param string $signature Digital signature as hex string
+	 * @return hex string Timestamped digital signature
+	 * @protected
+	 * @author M Hida
+	 * @since 6.6.2 (2023-05-25)
+	 */
```
This documentation is misleading since the function already existed in prior versions
could you update this to reflect that it already existed ?

hidasw replied:
probably remove the previous comment, right?

williamdes replied:
thanks, that's better

williamdes reviewed (dismissed):
Very nice !
When I find time I will do some polishing of the code indentation

hidasw committed:
Update tcpdf.php

hidasw commented:
@williamdes i think this will be merged. :)

williamdes commented:
> @williamdes i think this will be merged. :)

did you get some news ?

hidasw commented:
> > @williamdes i think this will be merged. :)
> 
> did you get some news ?

No, what news?

williamdes commented:
> No, what news?

None. So we still have to wait for more review by users that can merge this PR

williamdes commented on include/tcpdf_cmssignature.php:
```diff
@@ -0,0 +1,545 @@
+ * @version 1.0.001
+ * @author M Hida
+ */
+class tcpdf_cms_signature {
```
can this file have less code ?

hidasw replied:
Seems hard. It's better if there's another FOSS library we can use to append an existing PKCS7 structure or employ other methods, instead of parsing and appending PKCS7 data standalone. However, they are usually intended for general purposes with many functions that we don't even need here.

williamdes commented on include/tcpdf_cmssignature.php:
```diff
@@ -0,0 +1,545 @@
+	 * Catch error.
+	 * @public
+	 */
+	public string $errorMsg;
```
```suggestion
	public $errorMsg;
```

You need to remove them, old php compatibility is required

hidasw replied:
Glad to see you actively maintain this repo

hidasw committed:
Update include/tcpdf_cmssignature.php

Co-authored-by: William Desportes <williamdes@wdes.fr>

csalzano commented:
Hello. I need this feature. @hidasw are you willing to update your fork until this gets merged so my project can use the fork as a dependency?

hidasw commented:
@csalzano There are no updates planned at this time.
Still waiting for everyone's suggestions regarding what updates and features need to be made.
Currently it works quite well with existing tcpdf signatures.

evamtinez commented on include/tcpdf_cmssignature.php:
```diff
@@ -0,0 +1,545 @@
+	 * result value of pkcs7 EncryptedDigest.
+	 * @public
+	 */
+	public string $pkcs7_EncryptedDigest;
```
Typed properties are only allowed since PHP 7.4. As this library require "php": ">=5.3.0", this may cause problems in php versions under 7.4.

evamtinez commented on include/tcpdf_cmssignature.php:
```diff
@@ -0,0 +1,545 @@
+		curl_close($ch);
+
+		if($tsResponse) {
+			$header_size = curl_getinfo($ch, CURLINFO_HEADER_SIZE);
```
This code showed me an error because this is trying to access to curl info after doing curl_close. Can you check it?

Thank you.

hidasw replied:
yes curl_close($ch); should be moved to after curl_getinfo . maybe a php version compatibility issue, I think this has worked before (or I forgot) $tsResponse variable should save the curl_exec($ch) resource and keep it even if it has been closed.

evamtinez commented:
Hi @hidasw 

I would like to use this functionallity in my project and I identified a few lines that require adjustments.

I've just sent you some feedback on your code, but I'm unsure of the next steps. Should I perhaps suggest the changes via a pull request to your fork?

Thank you for your efforts.

hidasw commented:
@evamtinez  of course you are welcome

evamtinez committed:
Remove type on properties.

evamtinez committed:
Curl close was moved after curl_getinfo.

evamtinez committed:
Fix indentation

evamtinez commented:
> @evamtinez of course you are welcome

I created a PR in your fork https://github.com/hidasw/TCPDF/pull/2

Feel free to give feedback or to make changes @hidasw

hidasw committed:
Merge pull request #2 from evamtinez/main

Fix some compatibility issues with lower versions of PHP

williamdes commented on include/tcpdf_cmssignature.php:
```diff
@@ -0,0 +1,545 @@
+<?php
```
If this file could have a code formatting that would be great.
Else, it looks okay to me

hidasw replied:
code formatting.. like what?
Sorry, I do not understand what you mean. :)

williamdes commented on include/tcpdf_cmssignature.php:
```diff
@@ -0,0 +1,545 @@
+<?php
```
why `cms` signature ?
and not TSA ?

williamdes replied:
oh okay, good name then

hidasw committed:
Add TSA parameter

Use Apple TSA for example

hidasw commented:
> Hello. I need this feature. @hidasw are you willing to update your fork until this gets merged so my project can use the fork as a dependency?

see example_052.php on my fork

hidasw committed:
update to support LTV and TSA at once

hidasw committed:
Merge branch 'main' into main

hidasw commented:
The latest update has support for LTV.

hidasw committed:
significant improvement on signing process

- Signing process no longer need to create and read temporary file.
And no longer tempfile created by openssl_pkcs7_sign().
- Added option to set signature hash algorithm.

hidasw committed:
Merge branch 'main' of https://github.com/hidasw/TCPDF

hidasw committed:
combines several functions

hidasw committed:
Update example_052.php

hidasw committed:
Update tcpdf.php

hidasw committed:
Update tcpdf_cmssignature.php

hidasw committed:
Update example_052.php

hidasw committed:
Delete RootCATest.pem.crl

hidasw committed:
Delete RootCATest.der.crl

hidasw committed:
Delete Root CA Test.crt

hidasw committed:
Delete Root CA Test OCSP Signer.pem

hidasw committed:
Delete PDF User.pem

hidasw committed:
Create PDF Signing CA.crt

hidasw committed:
Create ocspTest.bat

hidasw committed:
Create longChain.pfx

hidasw committed:
Create longChain.pem

hidasw committed:
Create long.pfx

hidasw committed:
Create b6ce4782.0

hidasw committed:
Create 7262ea48.0

hidasw committed:
Create 71ab782e.0

hidasw committed:
Create 5a0ce691.0

hidasw committed:
Create 26a91b3f.0

evamtinez commented:
Hi @hidasw 

I would like to use this functionality in a production environment. Is it ready to be merged?

Thank you

evamtinez commented:
Hi @hidasw,

I have noticed that the code does not validate the TSA URL at any point, and passing an invalid URL does not display any error. It simply does not apply the timestamp to the signature. I wanted to ask if this is a functionality planned to be added to the MR.

Thanks

hidasw commented:
@evamtinez 
Last time I checked the script was running fine on php 7.4. I haven't had time to test thoroughly on php 8 or php 5.4. 
What version of PHP are you currently using?
For the display error problem, I can't display it in the output buffer because it will go into the PDF file (currently I haven't used the error handler in TCPDF).
To view the log, it is in the `'../signature.log'` file in the TCPDF directory. Can be set in the `tcpdf_cmssignature` class by setting `$writeLog` which is true by default.
To really make sure that there are no unhandled error notifications in the PDF file, please open the PDF file in a text editor and make sure there are no PHP error notifications in the PDF file.

evamtinez commented:
> @evamtinez Last time I checked the script was running fine on php 7.4. I haven't had time to test thoroughly on php 8 or php 5.4. What version of PHP are you currently using? For the display error problem, I can't display it in the output buffer because it will go into the PDF file (currently I haven't used the error handler in TCPDF). To view the log, it is in the `'../signature.log'` file in the TCPDF directory. Can be set in the `tcpdf_cmssignature` class by setting `$writeLog` which is true by default. To really make sure that there are no unhandled error notifications in the PDF file, please open the PDF file in a text editor and make sure there are no PHP error notifications in the PDF file.

I'm using PHP 7.2. 

I can see the log file correctly, but it would be helpful to get an error if the timestamp is invalid. For instance, if I set a timestamp to something like tsa.example.org (which doesn’t exist), the document appears to be signed correctly, and the process returns OK, but the PDF file doesn’t have the timestamp.

I’m handling this error in the class I'm using in my project, but I'm unsure if it needs to be directly integrated into the TCPDF library. In my implementation, before setting the timestamp, I call a validation function to check if the TSA is valid. However, this validation could be directly implemented in the setTimestamp function. What do you think?

Thanks

hidasw commented:
@evamtinez 
Generally tsa is optional, so it's probably better to ignore it if it fails than to abort the entire signature.

I think so.
Since first PR, I thought about providing an option to cancel the process if timestamp or LTV validation fails. But maybe this does not cancel the signature field/signature appearance that has been called, so it will provide null data in the signature which makes the signature invalid. I haven't tried it yet, maybe someday.

As a note, for the current version, if the validation process fails on one  of certificate chains, then this certificate chain up to the root CA is not embedded in the pdf. (to be fixed in the next update).
Thank's for suggestion.

hidasw commented on include/tcpdf_cmssignature.php:
```diff
@@ -0,0 +1,1847 @@
+          $time = $params[0]; //yymmddhhiiss
+          $oldTz = date_default_timezone_get();
+          date_default_timezone_set("UTC");
+          $time = date("ymdHis", $time);
```
plan to use strtotime or date class, but need to manage while converting 4 digits year to 2 digits. if year is 70, some app treat it as 1970 and other as 2070. also for time zone difference.

rsvitak commented:
I am testing this fork and after I added the curl_setopt($ch, CURLOPT_USERPWD, 'username:password'); in the sendReq function (tcpdf_cmssignature.php) it successfuly requested the specified TSA, but I don't think the TSR was added correctly (I am testing the final pdf file with pdfsig utility though).

However, I would like to ask you if you have idea how to add *ONLY* the the timestamp signature, not the ID signiture /certificate?

hidasw commented:
whats the mean "but I don't think the TSR was added correctly", is there a problem with tsresponse?
About timestamp only, I see that Adobe reader can do it, but I have never tried with php script. I also don't understand what the timestamp is for without a digital signature.

rsvitak commented:
I am sorry for the late reply, I am struggling with this topic, everything about PDF format and cryptography is new to me. What I am looking for perfectly show this thread: https://stackoverflow.com/questions/65807104/can-you-add-a-timestamped-no-tamper-proof-to-a-pdf-without-signing-it

I have a working TSA, I am able to create a valid TSQ of a whole PDF file, send it to the TSA and I receive the valid TSR. As I am generating PDF files in PHP (currently with DomPdf, but I believe I can switch to TCPDF) I am trying to get to the state the TSR is embedded into the PDF.

Currently I am able to embed the timestamp into the signiture, but as I have only a self generated certificate the acrobat says the validity of the signiture is unknown. Therefore my question is how to embed ONLY the RFC3161, i.e. /Type /DocTimeStamp object into the PDF.


I do appologize for chaotic and unprecise expressions but hopefully it's understandable what I mean.

rsvitak commented:
I was able to get access to an Adobe Acrobat, configure time stamp authority and sign a PDF document only with the DTS. Here is a screenshot how it's then displayed in Acrobat. This is what I need to achieve with TCPDF:
![dts](https://github.com/user-attachments/assets/3ea7b2ad-bdcb-4ff2-9c2d-3d512359337b)

rsvitak commented:
I was finally ably to achieve the same result as when adding the TSA timestamp to the document by modifying (locally) this fork.

I made a dirty solution - added support to "special" signiture certificate (signature_data['signcert']) "DTS-ONLY" which turns the signing procedure to add only the TSR response into the /Type /DocTimeStamp PDF section.

I modified these files:
tcpdf.php:
`7660c7660
<               if ($this->sign) {
---
>               if ($this->sign || !empty($this->signature_data_tsa)) {
13415a13416,13423
>               if ($this->signature_data['signcert']==='DTS-ONLY') {
>                       $out .= '<<'."\n".'/Filter /Adobe.PPKLite'."\n";
>                       $out .= '/SubFilter /ETSI.RFC3161'."\n";
>                       $out .= TCPDF_STATIC::$byterange_string."\n";
>                       $out .= '/Contents<'.str_repeat('0', $this->signature_max_length).'>'."\n";
>                       $out .= '/Type /DocTimeStamp'."\n";
>                       $out .= '/V 0'."\n";
>               } else {
13475a13484
>       }
`
include/tcpdf_cmssignature.php:
`425a429,430
> 
>     if ($this->signature_data['signcert']!=='DTS-ONLY') {
580a586,596
>    
>     } else {
>       $this->log .= ("info: DTS-ONLY starts ...\n");
>       if ($TSTInfo=self::createTimestamp($binaryData, $hashAlgorithm)) {
>          $this->log .= ("info: DTS-ONLY SUCCESS.\n");
>       } else {
>          $this->log .= ("info: DTS-ONLY FAILED!\n");
>       }
>       $pkcs7ContentInfo=$TSTInfo;
>     }
> 
`
I am signing my documents with this simple code now:
`            $pdf=new TCPDF();
            $pdf->setSignature('DTS-ONLY', file_get_contents('postsignum_tsa_tsu1.pem'), '', '', 1, []);
            $pdf->setTimeStamp(TSA_URL, TSA_USERNAME, TSA_PASSWORD);
            $pdf->loadHtml($html);
            $pdf->setSignatureAppearance(20,10,50,15);
            $output=$pdf->Output($filename,'S');
`
I get results like this:
![obrazek](https://github.com/user-attachments/assets/1b12386e-286d-4e42-8962-b69c17453ab5)

KruzstofWren commented:
> I was finally ably to achieve the same result as when adding the TSA timestamp to the document by modifying (locally) this fork.
> 
> I made a dirty solution - added support to "special" signiture certificate (signature_data['signcert']) "DTS-ONLY" which turns the signing procedure to adding just the TSR response into the /Type /DocTimeStamp PDF section.
> 
> I modified these files: tcpdf.php:
> 
> ```
> 7660c7660
> <               if ($this->sign) {
> ---
> >               if ($this->sign || !empty($this->signature_data_tsa)) {
> 13415a13416,13423
> >               if ($this->signature_data['signcert']==='DTS-ONLY') {
> >                       $out .= '<<'."\n".'/Filter /Adobe.PPKLite'."\n";
> >                       $out .= '/SubFilter /ETSI.RFC3161'."\n";
> >                       $out .= TCPDF_STATIC::$byterange_string."\n";
> >                       $out .= '/Contents<'.str_repeat('0', $this->signature_max_length).'>'."\n";
> >                       $out .= '/Type /DocTimeStamp'."\n";
> >                       $out .= '/V 0'."\n";
> >               } else {
> 13475a13484
> >       }
> ```
> 
> include/tcpdf_cmssignature.php:
> 
> ```
> 425a429,430
> > 
> >     if ($this->signature_data['signcert']!=='DTS-ONLY') {
> 580a586,596
> >    
> >     } else {
> >       $this->log .= ("info: DTS-ONLY starts ...\n");
> >       if ($TSTInfo=self::createTimestamp($binaryData, $hashAlgorithm)) {
> >          $this->log .= ("info: DTS-ONLY SUCCESS.\n");
> >       } else {
> >          $this->log .= ("info: DTS-ONLY FAILED!\n");
> >       }
> >       $pkcs7ContentInfo=$TSTInfo;
> >     }
> > 
> ```
> 
> I am signing my documents with this simple code now:
> 
> ```
> $pdf=new TCPDF();
> $pdf->setSignature('DTS-ONLY', file_get_contents('postsignum_tsa_tsu1.pem'), '', '', 1, []);
> $pdf->setTimeStamp(TSA_URL, TSA_USERNAME, TSA_PASSWORD);
> $pdf->loadHtml($html);
> $pdf->setSignatureAppearance(20,10,50,15);
> $output=$pdf->Output($filename,'S');
> ```
> 
> I get results like this which is exactly what I wanted to achieve: ![obrazek](https://private-user-images.githubusercontent.com/741308/362302468-1b12386e-286d-4e42-8962-b69c17453ab5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYwNTE5NDYsIm5iZiI6MTcyNjA1MTY0NiwicGF0aCI6Ii83NDEzMDgvMzYyMzAyNDY4LTFiMTIzODZlLTI4NmQtNGU0Mi04OTYyLWI2OWMxNzQ1M2FiNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwOTExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDkxMVQxMDQ3MjZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00NTUzY2IxNWZmNDU1ZjNmYmM2NzhmOGVkODE5ODRiN2JjNTNjYjEwYWI0MjdhODEzZDYyYmQzZTA0NjQ0MzBhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.p2pfitCOtenxZe1cK8iA6Gq1PoEKnZFgOMV5sZlGXKM)

Please, can I ask @rsvitak i see you use postsignum TSA. Please can you help me with URL to create on this. I cant go throw autenticatin process. Many thx...
$pdf->setTimeStamp('https://tsa.postsignum.cz:444/TSS/HttpTspServer/', TSA_USERNAME, TSA_PASSWORD); 
??
Many thx

rsvitak commented:
@KruzstofWren I had to update the list of changes I made to the @hidasw code in my comment above, thank you for tour post. This is neccessary to add to the sendReq function in the include/tcpdf_cmssignature.pdf:
``    curl_setopt($ch, CURLOPT_USERPWD, $this->signature_data_tsa['username'].':'.$this->signature_data_tsa['password']);
``

hidasw commented:
@rsvitak 
Good news, you can add support for timestamp only.
I don't know much about pdf structure.
Maybe in the future it would be nice to add an option for timestamp only.

KruzstofWren commented:
> @KruzstofWren I had to update the list of changes I made to the @hidasw code in my comment above, thank you for tour post. This is neccessary to add to the sendReq function in the include/tcpdf_cmssignature.pdf to make the code send the autentication data to the server: ` curl_setopt($ch, CURLOPT_USERPWD, $this->signature_data_tsa['username'].':'.$this->signature_data_tsa['password']);`

Many thx for reaction. But still, in logs i have this fail...

info      :   Timestamping process start...info:    sending TSA query to "https://tsa.postsignum.cz:444/TSS/HttpTspServer/"...
info      :     TSA query OK
info      :     Parsing Timestamp response...FAILED!

Something more i need edit?

hidasw commented:
> > @KruzstofWren I had to update the list of changes I made to the @hidasw code in my comment above, thank you for tour post. This is neccessary to add to the sendReq function in the include/tcpdf_cmssignature.pdf to make the code send the autentication data to the server: ` curl_setopt($ch, CURLOPT_USERPWD, $this->signature_data_tsa['username'].':'.$this->signature_data_tsa['password']);`
> 
> Many thx for reaction. But still, in logs i have this fail...
> 
> info : Timestamping process start...info: sending TSA query to "https://tsa.postsignum.cz:444/TSS/HttpTspServer/"... info : TSA query OK info : Parsing Timestamp response...FAILED!
> 
> Something more i need edit?

I can't test it since tsa uses authentication.
But I can analyze it if I get the tsa response.
You can save the response to a file.
In the sendReq() function before return:

      $h = fopen('tsaResp.der','w');
      fwrite($h, $body);
      fclose($h);

parse the response with openssl:
     ` openssl asn1parse -i -inform der -in tsaResp.der`

Let me know whats wrong.

rsvitak commented:
I have stored my modifications to https://github.com/rsvitak/TCPDF to avoid missunderstandings with description of my modifications in my post https://github.com/tecnickcom/TCPDF/pull/617#issuecomment-2315346099.

KruzstofWren commented:
> > > @KruzstofWren I had to update the list of changes I made to the @hidasw code in my comment above, thank you for tour post. This is neccessary to add to the sendReq function in the include/tcpdf_cmssignature.pdf to make the code send the autentication data to the server: ` curl_setopt($ch, CURLOPT_USERPWD, $this->signature_data_tsa['username'].':'.$this->signature_data_tsa['password']);`
> > 
> > 
> > Many thx for reaction. But still, in logs i have this fail...
> > info : Timestamping process start...info: sending TSA query to "https://tsa.postsignum.cz:444/TSS/HttpTspServer/"... info : TSA query OK info : Parsing Timestamp response...FAILED!
> > Something more i need edit?
> 
> I can't test it since tsa uses authentication. But I can analyze it if I get the tsa response. You can save the response to a file. In the sendReq() function before return:
> 
> ```
>   $h = fopen('tsaResp.der','w');
>   fwrite($h, $body);
>   fclose($h);
> ```
> 
> parse the response with openssl: ` openssl asn1parse -i -inform der -in tsaResp.der`
> 
> Let me know whats wrong.

All is working great now. Only mistake is on my side....right URL for https://www.postsignum.cz is: https://www3.postsignum.cz/TSS/TSS_user/
Many thx to all, your are the best!

KruzstofWren commented:
If i may, last question. Do you think is much diference bettwen SIGN or CERTIFICATE document? Images are down  ... I dont see any diference. Both document have timestamp...

![obrazek](https://github.com/user-attachments/assets/d55a1a53-ef6a-443b-b9b9-5f5d169e3e34)

![obrazek](https://github.com/user-attachments/assets/ef85b421-6942-4637-b5e5-58e445c60c63)

rsvitak commented:
@KruzstofWren I am not very sure in this topic, but the difference could be in the authority that proves the "stamp"? Maybe if you could paste the "TCPDF" commands used to create the document, or if you can provide the final PDF files, it would be easier to tell more...

nicolaasuni commented:
Hi,
A lot of good work has been put into this feature that I will consider for the new tc-lib-pdf library instead.
I am trying to not accept any new feature for TCPDF so I can deprecate it sooner.

williamdes commented:
Positive test feedback received on https://github.com/tecnickcom/TCPDF/issues/523#issuecomment-3652117405 !

nicolaasuni commented:
Thank you very much, @hidasw, for this remarkable piece of work - building RFC 3161 timestamping support from scratch, including a custom ASN.1 parser and CMS signature class, is a significant undertaking and the community engagement this PR has attracted over more than two years speaks for itself.

TCPDF 6 is now **deprecated** and no further changes will be merged into this repository. Please see the pinned issue [Thank You - and the Future of TCPDF](https://github.com/tecnickcom/TCPDF/issues/867) for the full explanation and the path forward.

As you may have seen from the `consider_for_new_lib` label, this feature was already earmarked for **[tc-lib-pdf](https://github.com/tecnickcom/tc-lib-pdf)**, which already ships working TSA timestamp and LTV support (see examples E007-E009). I would warmly encourage you to open a PR or discussion there if there are aspects of your implementation that are not yet covered.

I am leaving this PR open in case you or anyone else wishes to continue the discussion, but it is unlikely that this PR will ever be merged.

Thank you again for your extraordinary patience and generosity.
